File: tests/fixtures/safe/sanitised.java
// FIXTURE: properly sanitised flows, Java.
// EXPECT-NONE
import java.sql.*;
import javax.servlet.http.HttpServletRequest;

public class SafeTaintFixture {
    // A number cannot carry a quote.
    public ResultSet numeric(HttpServletRequest request, Statement stmt) throws SQLException {
        int id = Integer.parseInt(request.getParameter("id"));
        return stmt.executeQuery("SELECT * FROM users WHERE id = " + id);
    }

    // Parameterised: the value is never part of the instruction.
    public ResultSet parameterised(HttpServletRequest request, Connection conn) throws SQLException {
        PreparedStatement ps = conn.prepareStatement("SELECT * FROM users WHERE id = ?");
        ps.setString(1, request.getParameter("id"));
        return ps.executeQuery();
    }
}

File: tests/fixtures/safe/sanitised.go
// FIXTURE: properly sanitised flows, Go.
// EXPECT-NONE
package fixtures

import (
	"database/sql"
	"fmt"
	"net/http"
	"os/exec"
	"strconv"
)

func numeric(r *http.Request, db *sql.DB) {
	id, _ := strconv.Atoi(r.FormValue("id"))
	db.Query("SELECT * FROM users WHERE id = " + fmt.Sprint(id))
}

func parameterised(r *http.Request, db *sql.DB) {
	db.Query("SELECT * FROM users WHERE id = $1", r.FormValue("id"))
}

// Arguments as a list: no shell parses them.
func noShell(r *http.Request) error {
	return exec.Command("ping", "-c", "1", r.FormValue("host")).Run()
}
